src/DiffResults.jl

```diff
@@ -79,7 +79,7 @@ shape information. If you want to allocate storage yourself, use the `DiffResult
 constructor instead.
 """
 JacobianResult(x::AbstractArray) = DiffResult(similar(x), similar(x, length(x), length(x)))
-JacobianResult(x::SArray{<:Any,T,<:Any,L}) where {T,L} = DiffResult(x, zeros(SMatrix{L,L,T}))
+JacobianResult(x::StaticArray) = DiffResult(x, zeros(StaticArrays.similar_type(typeof(x), Size(length(x),length(x)))))
```

Comment:
will the `length(x)` here cause the size parameter of the resulting array to be uninferrable, or does constant prop take care of that for us?

Reply:
Constant prop seems to take care of it (as I assumed without worrying much :smile:):
```julia
using DiffResults, StaticArrays, Test

v = SVector([1, 2.5, 4])
@inferred DiffResults.JacobianResult(v)
@code_warntype DiffResults.JacobianResult(v)
```
outputs
```julia
Body::DiffResults.ImmutableDiffResult{1,SArray{Tuple{3},Float64,1,3},Tuple{SArray{Tuple{3,3},Float64,2,9}}}
1 ─ %1 = %new(DiffResults.ImmutableDiffResult{1,SArray{Tuple{3},Float64,1,3},Tuple{SArray{Tuple{3,3},Float64,2,9}}}, x, ([0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0],))::DiffResults.ImmutableDiffResult{1,SArray{Tuple{3},Float64,1,3},Tuple{SArray{Tuple{3,3},Float64,2,9}}}
└──      return %1
```